crates/datafusion-optimizer-rules/src/logical_plan/duckdb/aggregate_pushdown.rs

```diff
@@ -0,0 +1,618 @@
+    fn try_percolate_marker_node(plan: &LogicalPlan) -> Result<Option<LogicalPlan>> {
+        let with_erased_marker = plan.clone().transform_down(|p| {
+            if p.inputs().len() > 1 || matches!(p, LogicalPlan::Analyze(_)) {
+                return Ok(Transformed::new(p, false, TreeNodeRecursion::Stop));
+            }
+
+            let LogicalPlan::Extension(ref ext) = p else {
+                return Ok(Transformed::no(p));
+            };
+
+            let Some(marker) = ext
+                .node
+                .as_any()
+                .downcast_ref::<DuckDBAggregatePushdownNode>()
+            else {
+                return Ok(Transformed::no(p));
+            };
+
+            Ok(Transformed::new(
+                marker.input_plan.clone(),
+                true,
+                TreeNodeRecursion::Jump,
+            ))
+        })?;
+
+        if with_erased_marker.transformed {
+            Ok(Some(LogicalPlan::Extension(Extension {
+                node: DuckDBAggregatePushdownNode::new(with_erased_marker.data),
+            })))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl OptimizerRule for DuckDBAggregateLogicalPushdown {
+    fn name(&self) -> &'static str {
+        "DuckDBAggregatePushdownOptimizerRule"
+    }
+
+    // This rule does its own recursion
+    fn apply_order(&self) -> Option<ApplyOrder> {
+        None
+    }
+
+    fn rewrite(
+        &self,
+        plan: LogicalPlan,
+        _config: &dyn OptimizerConfig,
+    ) -> Result<Transformed<LogicalPlan>, DataFusionError> {
+        // Mark all eligible nodes for DuckDB agg pushdown
+        let maybe_marked_agg = plan.transform_down(|p| {
+            if let LogicalPlan::Extension(ext) = &p
+                && concrete!(ext.node, DuckDBAggregatePushdownNode).is_some()
+            {
+                return Ok(Transformed::new(p, false, TreeNodeRecursion::Jump));
+            }
+
+            if let Some(marked_for_pushdown) = Self::try_mark_pushdown(&p)? {
+                Ok(Transformed::new(
+                    marked_for_pushdown,
+                    true,
+                    TreeNodeRecursion::Jump,
+                ))
+            } else {
+                Ok(Transformed::no(p))
+            }
+        })?;
+
+        // If we didn't rewrite, bail out early
+        if !maybe_marked_agg.transformed {
+            return Ok(maybe_marked_agg);
+        }
+
+        // Try to push as much of the logical plan under the pushdown marker as possible. We
+        // do this in two steps since the previous only operates on aggregate nodes (it is not
+        // possible to walk up at the point in time of rewriting), and trying to account for all
+        // invariants in one steps is difficult to follow
+        let rewritten_plan = maybe_marked_agg.data;
+        rewritten_plan.transform_down(|p| {
+            if let Some(percolated) = Self::try_percolate_marker_node(&p)? {
```

Comment:
Same here, if you returned Transformed from `try_percolate_marker_node` then you could avoid the clone